curp/src/server/curp_node.rs

```diff
@@ -173,42 +173,101 @@ impl<C: 'static + Command> CurpNode<C> {
 
         let mut ticker = tokio::time::interval(heartbeat_interval);
         ticker.set_missed_tick_behavior(MissedTickBehavior::Delay);
-        #[allow(clippy::integer_arithmetic, unused_attributes)] // tokio internal triggered
         loop {
             let _now = ticker.tick().await;
-            let action = curp.tick();
-            match action {
-                TickAction::Heartbeat(hbs) => {
-                    Self::bcast_heartbeats(Arc::clone(&curp), &connects, hbs).await;
-                }
-                TickAction::Votes(votes) => {
-                    Self::bcast_votes(Arc::clone(&curp), &connects, votes).await;
-                }
-                TickAction::Nothing => {}
+            if let Some(vote) = curp.tick_election() {
+                Self::bcast_vote(curp.as_ref(), &connects, vote).await;
             }
         }
     }
 
-    /// Background leader calibrate followers
-    async fn calibrate_task(
+    /// Sync task is responsible for syncing a follower
+    async fn sync_task(
         curp: Arc<RawCurp<C>>,
-        connects: HashMap<ServerId, Arc<impl ConnectApi>>,
-        mut calibrate_rx: mpsc::UnboundedReceiver<ServerId>,
+        connect: Arc<impl ConnectApi>,
+        sync_event: Arc<Event>,
     ) {
-        let mut handlers: HashMap<ServerId, JoinHandle<()>> = HashMap::new();
-        while let Some(follower_id) = calibrate_rx.recv().await {
-            if handlers
-                .get(&follower_id)
-                .map_or(false, |hd| !hd.is_finished())
-            {
+        let leader_event = curp.leader_event();
+        loop {
+            if !curp.is_leader() {
+                leader_event.listen().await;
+            }
+            Self::_sync_task(
+                Arc::clone(&curp),
+                Arc::clone(&connect),
+                Arc::clone(&sync_event),
+            )
+            .await;
+        }
+    }
+
+    /// real sync task, will stop if self is no long the leader
+    async fn _sync_task(
+        curp: Arc<RawCurp<C>>,
+        connect: Arc<impl ConnectApi>,
+        sync_event: Arc<Event>,
+    ) {
+        let mut hb_opt = false;
+        let mut ticker = tokio::time::interval(curp.cfg().heartbeat_interval);
+        ticker.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        let id = connect.id();
+        let rpc_timeout = curp.cfg().rpc_timeout;
+
+        #[allow(clippy::integer_arithmetic)] // tokio select internal triggered
+        loop {
+            // a sync is either triggered by an heartbeat timeout event or when new log entries arrive
+            tokio::select! {
+                _now = ticker.tick() => {
+                    if hb_opt {
+                        hb_opt = false;
+                        continue;
+                    }
+                }
+                // TODO: use a batch timer
+                _ = sync_event.listen() => {}
```

Comment:
The sync_event is only used in the `handle_vote_response` function. Shall we give it a more meaningful name?

Reply:
It may miss some event during the loop as an event may be dropped if there's no listener waiting.

Reply:
> The sync_event is only used in the `handle_vote_response` function. Shall we give it a more meaningful name?

It is also used in `handle_propose`. I think this line of code is miss deleted when rebasing and now I've added it. `sync_event` means the leader should trigger a sync with the follower.

Reply:
> It may miss some event during the loop as an event may be dropped if there's no listener waiting.

Fixed.